File: web/app/lib/knowhow/queries.ts
import "server-only";
import { cache } from "react";
import { gql } from "./client";
import type { Vault, Document, DocumentSummary } from "./types";

export const getVaults = cache(async (): Promise<Vault[]> => {
  const data = await gql<{ vaults: Vault[] }>(`
    query {
      vaults {
        id
        name
        description
      }
    }
  `);

  return data.vaults;
});

export const getVaultDocuments = cache(
  async (vaultId: string): Promise<DocumentSummary[]> => {
    const data = await gql<{ vault: { documents: DocumentSummary[] } | null }>(
      `
    query ($id: ID!) {
      vault(id: $id) {
        documents {
          id
          vaultId
          path
          title
          labels
          docType
          createdAt
          updatedAt
        }
      }
    }
  `,
      { id: vaultId },
    );

    return data.vault?.documents ?? [];
  },
);

export type FolderRecord = {
  path: string;
};

export const getVaultFolders = cache(
  async (vaultId: string): Promise<FolderRecord[]> => {
    const data = await gql<{ vault: { folders: FolderRecord[] } | null }>(
      `
    query ($id: ID!) {
      vault(id: $id) {
        folders {
          path
        }
      }
    }
  `,
      { id: vaultId },
    );

    return data.vault?.folders ?? [];
  },
);

export async function getDocument(
  vaultId: string,
  path: string,
): Promise<Document | null> {
  const data = await gql<{ document: Document | null }>(
    `
    query ($vaultId: ID!, $path: String!) {
      document(vaultId: $vaultId, path: $path) {
        id
        vaultId
        path
        title
        content
        contentBody
        labels
        docType
        createdAt
        updatedAt
        wikiLinks {
          id
          fromDocId
          toDocId
          rawTarget
          resolved
        }
        backlinks {
          id
          fromDocId
          toDocId
          rawTarget
          resolved
        }
      }
    }
  `,
    { vaultId, path },
  );

  return data.document;
}

File: ios/Networking/Queries.swift
import Foundation

/// GraphQL query string constants matching the Knowhow server schema.
enum Queries {
    static let me = """
    query Me {
        me {
            user {
                id
                name
                email
                createdAt
            }
            vaultRoles {
                vaultId
                role
            }
        }
    }
    """

    static let vaults = """
    query Vaults {
        vaults {
            id
            name
            description
            createdAt
            updatedAt
            labels
        }
    }
    """

    static let vault = """
    query Vault($id: ID!, $folder: String) {
        vault(id: $id) {
            id
            name
            description
            createdAt
            updatedAt
            labels
            folders(parent: $folder) {
                id
                path
                name
                createdAt
            }
            documents(folder: $folder) {
                id
                path
                title
                labels
                docType
                updatedAt
            }
        }
    }
    """

    static let document = """
    query Document($vaultId: ID!, $path: String!) {
        document(vaultId: $vaultId, path: $path) {
            id
            vaultId
            path
            title
            content
            contentBody
            labels
            docType
            source
            createdAt
            updatedAt
        }
    }
    """

    static let documentById = """
    query DocumentById($id: ID!) {
        documentById(id: $id) {
            id
            vaultId
            path
            title
            content
            contentBody
            labels
            docType
            source
            createdAt
            updatedAt
        }
    }
    """

    static let search = """
    query Search($input: SearchInput!) {
        search(input: $input) {
            documentId
            path
            title
            labels
            docType
            score
            matchedChunks {
                snippet
                headingPath
                position
                score
            }
        }
    }
    """
}
